File: 프로그래머스/1/42889. 실패율/실패율.js
function solution(N, stages) {
    
    const stageMap = {}
    
    for(let i = 1; i <= N; i++){
        stageMap[i] = {
            'thru': 0,
            'stuck': 0
        }
    }
    
    stages.forEach(stage => {
        for (let i = 1; i <= Math.min(stage, N); i++) {
            if (i === stage && stage <= N) stageMap[i].stuck++;
              stageMap[i].thru++;
        }
    })
    
    
    const sort = Object.entries(stageMap).map(stage => [stage[0], stage[1].thru === 0 ? 0 : stage[1].stuck / stage[1].thru]).sort((a, b) => b[1] - a[1]).map(i => Number(i[0]));
    
    
    return sort
}

File: 프로그래머스/1/42889. 실패율/실패율.py
def solution(N, stages):
    stage_dict = []
    num_of_players = len(stages)
    
    for i in range(N + 1):
        stage_dict.append(0)
        
    for s in stages:
        stage_dict[s - 1] += 1
    
    failure = []
    reach = 0
    for i in range(N + 1, 0, -1):
        if i == N + 1:
            reach = stage_dict[i - 1]
            continue
        else:
            reach += stage_dict[i - 1]
            
            if reach == 0:
                failure.append(0)
            else:
                failure.append(stage_dict[i - 1] / reach)
        
    failure.reverse()
    
    # (스테이지 번호, 실패율)
    result = [(i + 1, failure[i]) for i in range(N)]

    # 실패율 기준 내림차순 정렬 (값 기준)
    result.sort(key=lambda x: x[1], reverse=True)

    # 스테이지만 추출
    return [stage for stage, _ in result]
        